src/sources/splunk_hec/mod.rs

```diff
@@ -682,57 +701,47 @@ struct EventIterator<'de, R: JsonRead<'de>> {
     log_namespace: LogNamespace,
     /// handle to EventsReceived registry
     events_received: Registered<EventsReceived>,
+    /// Whether to store the HEC token in the log event metadata
+    store_hec_token: bool,
+}
+
+pub struct RequestMeta {
+    token: Option<String>,
+    remote: Option<SocketAddr>,
+    remote_addr: Option<String>,
 }
 
 /// Intermediate struct to generate an `EventIterator`
 struct EventIteratorGenerator<'de, R: JsonRead<'de>> {
     deserializer: serde_json::StreamDeserializer<'de, R, JsonValue>,
     channel: Option<String>,
     batch: Option<BatchNotifier>,
-    token: Option<Arc<str>>,
     log_namespace: LogNamespace,
     events_received: Registered<EventsReceived>,
-    remote: Option<SocketAddr>,
-    remote_addr: Option<String>,
+    meta: RequestMeta,
+    store_hec_token: bool,
 }
 
-impl<'de, R: JsonRead<'de>> From<EventIteratorGenerator<'de, R>> for EventIterator<'de, R> {
+impl<'de, R: JsonRead<'de>, E: Extractor> From<EventIteratorGenerator<'de, R>>
+    for EventIterator<'de, R, E>
+{
     fn from(f: EventIteratorGenerator<'de, R>) -> Self {
         Self {
             deserializer: f.deserializer,
             events: 0,
             channel: f.channel.map(Value::from),
             time: Time::Now(Utc::now()),
-            extractors: [
-                // Extract the host field with the given priority:
-                // 1. The host field is present in the event payload
-                // 2. The x-forwarded-for header is present in the incoming request
-                // 3. Use the `remote`: SocketAddr value provided by warp
-                DefaultExtractor::new_with(
-                    "host",
-                    log_schema().host_key().cloned().into(),
-                    f.remote_addr
-                        .or_else(|| f.remote.map(|addr| addr.to_string()))
-                        .map(Value::from),
-                    f.log_namespace,
-                ),
-                DefaultExtractor::new("index", OptionalValuePath::new(INDEX), f.log_namespace),
-                DefaultExtractor::new("source", OptionalValuePath::new(SOURCE), f.log_namespace),
-                DefaultExtractor::new(
-                    "sourcetype",
-                    OptionalValuePath::new(SOURCETYPE),
-                    f.log_namespace,
-                ),
-            ],
+            token: f.meta.token.clone().map(Into::into),
+            extractors: E::new(f.meta, f.log_namespace),
```

Comment:
Could this be moved into `EventIteratorGenerator` instead of `new` being called here? How far up the call chain could that go?